EPPs/common.py

```diff
@@ -144,6 +153,86 @@ def run(self):
             self.error('Stack trace below:\n' + stacktrace)
             raise e
 
+    @cached_property
+    def input_container_names(self):
+        """The name of containers from input artifacts. """
+        containers = set()
+        for art in self.artifacts:
+            # Check to see if artifact has a container before retrieving the container.
+            # Artifacts that are not samples will not have containers.
+            if art.container:
+                containers.add(art.container.name)
+        return list(containers)
+
+    @cached_property
+    def output_container_names(self):
+        """The name of containers from output artifacts"""
+        containers = set()
+        for art in self.output_artifacts:
+            # Check to see if artifact has a container before retrieving the container.
+            # Artifacts that are not samples will not have containers.
+            if art.container:
+                containers.add(art.container.name)
+        return list(containers)
+
+    def _validate_step(self):
+        """Perform the Step EPP's validations when required.
+        All validations are optionals and will require sub classes to set the corresponding varaibles:
+
+         - _max_nb_input_containers: set the maximum number of input container of the step.
+         - _max_nb_output_containers: set the maximum number of output container of the step.
+         - _max_nb_input: set the maximum number of input of the step.
+         - _nb_analyte_per_input: set the number of replicates (Analytes) required for the step.
+         - _nb_resfile_per_input: set the number of replicates (ResultFiles) required for the step.
+         - _max_nb_project: set the maximum number of project in the step.
+        """
+        if self._max_nb_input_containers is not None:
+            # check the number of input containers
+            if len(self.input_container_names) > self._max_nb_input_containers:
+                raise InvalidStepError(
+                    message='Maximum number of input container is %s. There are %s input container in the step.' % (
```

Comment:
[This requested change](https://github.com/EdinburghGenomics/clarity_scripts/pull/69#discussion_r232696073) is still not fully implemented.

Reply:
`InvalidStepError` now inherit from `Exception` and the parameter `message` is required to make it mandatory and explicit to provide a message to be display.

Reply:
As discussed, the single argument to the constructor makes sure only one argument is passed. It doesn't call the superconstructor, though - `super().__init__` call to be added.